### ISSUE TITLE
Feature request for consideration: build() on demand

### ISSUE THREAD

**Issue by wowczarek** [json-editor/json-editor#913]
Not an issue as such - raising this to see if the below is feasible. if(tlDR) goto tlDR, albeit still a long read; apologies.

I have been using json-editor for nearly two years now in a work project that's now been put into production and I have actual users breaking / using it. I've made tons of changes to 1.4.0 so I need to catch up with git latest and maybe submit some PRs. This will come later. 

I am working with a large schema that uses a lot of watchers, but even without that it's simply big and filled with lots of data. The schema itself is heading towards 50kB. I knew I may have to break the data into separate editors and load on demand eventually, but scale effect kicked in before I had the time to scratch my head.  Yes, I've made a very lazy design decision; I'm a shit software architect like that. 10+ levels of depth, lots of arrays of objects with array-of-object properties, etc, etc. I would happily share the schema as it's a good torture test, but I first need to write a tool that obfuscates it.

tlDR:

Very large schema, lots of data (not megabytes yet but approaching), 10+ levels I think, arrays at top level multiplying most of everything, lots of watchers. Editor is starting to be painfully slow to load. Yes, there have been performance improvements since 1.4.0 such as DOM vs innerHTML(), which I also did, but it's the sheer effort of building all those child editors that is the bottleneck; I've profiled it ad nauseam.

I noticed a max_depth or max_level option or similar in the latest versions but have not tested it yet, and it doesn't seem like this flat approach would get me where I want. What I'm considering is build()/preBuild() on uncollapsing child editors only. Most of the children I have are collapsed at start. So values and events would be maintained in memory, but no operations would happen on the view side, visible or not, unless something is uncollapsed (made visible first time), which is when visible editors are built and then shown, with a callback for a "start" and "end" event, so the UI can trigger a spinner or some other action while this happens. Any triggers that update titles, labels, etc, would happen in memory, but none of the rendering and building would be touched unless something is either visible or already built.

Questions:

- has this approach been suggested or considered before? I think I saw an issue about handling large data sets...
- have I missed something and can the editor be built headless already somehow to speed things up? Note that I not only set the initial data, but also load a completely new set of data sometimes, because the system in question can load different data sets.

Thanks!

**Comment by schmunk42** [comment]
> has this approach been suggested or considered before? I think I saw an issue about handling large data sets...

Don't think so. `json-editor` originated in a time when there was no (or no wide usage of) shadow DOM.

> have I missed something and can the editor be built headless already somehow to speed things up? Note that I not only set the initial data, but also load a completely new set of data sometimes, because the system in question can load different data sets.

Not as I know. 

CC: @germanbisurgi @amk0

**Comment by wowczarek** [comment]
All right - I'll take a stab at it. I'll have to move to master and port all the cruft I've added since first. I'll probably post some PRs as some of the stuff I added may be useful, like:

- setting a "data-schemapath" attribute for inputs or otherwise holders with the editor's path, so the relevant item can be focused on and such, plus code to uncollapse everything in the way from top to the item.
- a helper for automatically incrementing ID fields. Yes, we have uuids, but these are numeric. Debatable.
- a new callback for enums: transform. This is so watched data can be passed to a function that constructs the title,value array the way the user sees fit
- other bug fixes, some have not been fixed in master yet I think, a notable one is a fix for enums resetting value if watched data is not their direct ancestor and is still being loaded.

Will be in touch, eventually :)
